src/Dapr.Workflow/WorkflowContext.cs

```diff
@@ -0,0 +1,78 @@
+// ------------------------------------------------------------------------
+// Copyright 2022 The Dapr Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//     http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ------------------------------------------------------------------------
+
+namespace Dapr.Workflow
+{
+    using System;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.DurableTask;
+
+    /// <summary>
+    /// Context object used by workflow implementations to perform actions such as scheduling activities, durable timers, waiting for
+    /// external events, and for getting basic information about the current workflow instance.
+    /// </summary>
+    public class WorkflowContext
+    {
+        readonly TaskOrchestrationContext innerContext;
+
+        internal WorkflowContext(TaskOrchestrationContext innerContext)
+        {
+            this.innerContext = innerContext ?? throw new ArgumentNullException(nameof(innerContext));
+        }
+
+        /// <summary>
+        /// Gets the name of the current workflow.
+        /// </summary>
+        public TaskName Name => this.innerContext.Name;
+
+        /// <summary>
+        /// Gets the instance ID of the current workflow.
+        /// </summary>
+        public string InstanceId => this.innerContext.InstanceId;
+
+        /// <summary>
+        /// Gets the current workflow time in UTC.
+        /// </summary>
+        public DateTime CurrentUtcDateTime => this.innerContext.CurrentUtcDateTime;
+
+        /// <summary>
+        /// Assigns a custom status value to the current workflow.
+        /// </summary>
+        public void SetCustomStatus(object? customStatus) => this.innerContext.SetCustomStatus(customStatus);
+
+        /// <summary>
+        /// Creates a durable timer that expires after the specified delay.
+        /// </summary>
+        public Task CreateTimer(TimeSpan delay, CancellationToken cancellationToken = default)
+        {
+            return this.innerContext.CreateTimer(delay, cancellationToken);
+        }
+
+        /// <summary>
+        /// Waits for an event to be raised with name <paramref name="eventName"/> and returns the event data.
+        /// </summary>
+        public Task<T> WaitForExternalEventAsync<T>(string eventName, TimeSpan timeout)
+        {
+            return this.innerContext.WaitForExternalEvent<T>(eventName, timeout);
```

Comment:
Does this inner context not have an overload that accepts a cancellation token?  (I expect not, given the timeout, but just checking.)

Reply:
No, cancellation tokens would make it easy for developers to do some wrong things with workflows (for example, you wouldn't want to cancel a durable workflow operation simply because the app host is shutting down), so we opted for a design that instead incorporated timeouts via TimeSpan.

Reply:
Yes, I'm primarily worried that users might use it incorrectly since this is not a normal in-memory operation. It's very common in .NET to flow cancellation tokens to handle app shutdown events and it could result in unexpected behavior if developers blindly try to flow cancellation tokens to workflow context APIs. But I see your point that users may want to cancel for reasons other than timeout. I suggest we defer on this for now since the underlying API already doesn't support cancellation tokens (but I'm pinging the primary maintainer of this SDK to see if we should consider adding support for it later).

Reply:
Ugh, sorry, I was just corrected. We _do_ have an overload in the Durable Task SDK that accepts a cancellation token. The `TimeSpan` overload was mainly added for convenience. Since there's precedence for this, we can add a `WaitForExternalEventAsync` overload that takes a cancellation token if that would be preferred.

Reply:
Do you folks think it would be worth also opening an issue on this SDK to keep this registered and so we revisit this in the future? @cgillum, is there a GitHub issue to link here or in this possible issue?

Reply:
Not sure if our replies raced with each other, but the Durable Task SDK already has support for cancellation tokens here. It's up to use if we want to expose it. We can either do that now, open an issue to add support later, or just wait for customer feedback to see if they actually need/want this.

Reply:
I am exposing it in ad7ade292b2a495500da2f151b0f448e3c31fa53. Resolving this for now. Reopen if required.